### PR TITLE
spot_instance_request: fixed network interfaces configuration

### PR DESCRIPTION
**Pull request by hrach**
Fixes configuration public ip for spot_instance_request. Originally hashicorp/terraform#13736 & bug hashicorp/terraform#9575.

### CODE REVIEW

**Comment by hrach** [comment]
Hm, I can't see how it the test failure related to this PR. Previously it has passed test ok.

**Comment by radeksimko** [comment]
Hi @hrach 
I think the failure may be caused by wrong path. The right path begins with `aws/` whereas in your diff you're moving that file to `providers/aws/...`. 😉

**Comment by hrach** [comment]
@radeksimko thanks a lot. I have wrongly modified the patch; fixed now.

File: aws/resource_aws_spot_instance_request.go [hunk 1]
```diff
@@ -109,21 +109,14 @@ func resourceAwsSpotInstanceRequestCreate(d *schema.ResourceData, meta interface
...
 	if v, ok := d.GetOk("block_duration_minutes"); ok {
 		spotOpts.BlockDurationMinutes = aws.Int64(int64(v.(int)))
 	}
 
-	// If the instance is configured with a Network Interface (a subnet, has
-	// public IP, etc), then the instanceOpts.SecurityGroupIds and SubnetId will
-	// be nil
-	if len(instanceOpts.NetworkInterfaces) > 0 {
-		spotOpts.LaunchSpecification.SecurityGroupIds = instanceOpts.NetworkInterfaces[0].Groups
-		spotOpts.LaunchSpecification.SubnetId = instanceOpts.NetworkInterfaces[0].SubnetId
```
Comment (radeksimko) [inline aws/resource_aws_spot_instance_request.go]: Is there any reason we can't just add a this line here?

```go
spotOpts.LaunchSpecification.NetworkInterfaces = instanceOpts.NetworkInterfaces
```
We don't seem to have a test covering the case in the comment here, but I'd be worried that removing the logic may break some unrelated edge cases... 🤔

**Comment by hrach** [comment]
Ok, repushed.

- I've left my implementation. It seems correct to me. The pull request with the discussed logic is https://github.com/hashicorp/terraform/pull/3940. It also contains tests, so the if is/was tested. Since the method RequestSpotLaunchSpecification is from the official aws sdk, I think they may have also fixed and it should/may return correctly initizialized object.
- Added tests have passed:

```
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccAWSSpotInstanceRequest_SubnetAndSGAndPublicIpAddress -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSSpotInstanceRequest_SubnetAndSGAndPublicIpAddress
--- PASS: TestAccAWSSpotInstanceRequest_SubnetAndSGAndPublicIpAddress (151.46s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       151.509s
```

I'm definitely open to fix it as you wish, though, I have no idea how to make further inspection of aws-sdk.

**Comment by hrach** [comment]
Well, the bad thing is, that the test also pass when I do not apply the fix :-O So probably I would need some guidance what and how should I test the real instance setting.

**Comment by radeksimko** [comment]
> I think they may have also fixed and it should/may return correctly initizialized object.

That's a good point, although given the structure of the AWS SDK and the way it works it would has to be an API (server side) fix, because the SDK structures are pretty much all generated based on API JSON/XML structures. There's rarely any (if any) custom parsing logic. Server-side bugfix is certainly what may have happened though 👍 

I don't have too strong opinion on this, so as long as @catsby is happy as the original author of the linked PR, we're all happy. 😃 

> Well, the bad thing is, that the test also pass when I do not apply the fix

I think the reason the test is passing is because we're launching the instance into a subnet with auto-assignment enabled, as mentioned in [this comment](https://github.com/hashicorp/terraform/issues/9575#issuecomment-256499892).

If you set `map_public_ip_on_launch = false` in the test config you should see a failure (without the bugfix).

It's not necessary, but if you want you can also make the test more thorough by checking API response, e.g. adding a condition to `testAccCheckAWSSpotInstanceRequest_InstanceAttributes` or adding a new check for the spot request:

```
testAccCheckAWSSpotInstanceRequest_attributes(&sir),
```
where you can check contents of `sir.LaunchSpecification.NetworkInterfaces`.

As a side note it would be nice to separate this test instead of modifying an existing one so we can cover that specific use-case in isolation from others. Copy-pasting is totally aok in test cases.

**Comment by catsby** [comment]
Merged in https://github.com/terraform-providers/terraform-provider-aws/pull/1070 , thanks!
